jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcher.kt

```diff
@@ -0,0 +1,168 @@
+        return submit {
+            val res = supplier.invoke()
+            asyncOperationCompleted()
+            res
+        }.await()
+    }
+
+    /**
+     * Submits new operation to task queue.
+     *
+     *
+     * WARNING: task should not be long-running operation
+     * and should not block processing thread.
+     *
+     * @param supplier task to execute and retrieve result
+     * @return feature which represent result of task execution
+     */
+    private suspend fun <T> submit(supplier: suspend () -> T): Deferred<T> {
+        val state = state.get()
+        return if (state != State.RUNNING) {
+            throw RejectedExecutionException(
+                    "RateLimiterDispatcher [$name] is in '$state' state"
+            )
+        } else {
+            waitingScope.async {
+                processTask(supplier)
+            }
+        }
+    }
+
+    fun updateRate(rate: Int) {
+        rateLimiter.updateRate(rate)
+    }
+
+    override fun close() {
+        var stateUpdated = state.compareAndSet(State.RUNNING, State.SHUTTING_DOWN)
+        if (!stateUpdated) {
+            logger.info("Close called on RateLimitedDispatcher [{}] with state [{}]", name, state.get())
+            return
+        }
+
+        if (!waitingScope.waitChildrenAndCancel()) {
+            logger.warn("Timeout while waiting for coroutines finishing")
+            throw RejectedExecutionException("RateLimitedDispatcher [$name] get timeout while waiting for coroutines finishing")
+        }
+        stateUpdated = state.compareAndSet(State.SHUTTING_DOWN, State.TERMINATE)
+        if (!stateUpdated) {
+            logger.error(
+                "Can't set [TERMINATE] state to RateLimitedDispatcher [{}] in [{}] state",
+                name, state.get()
+            )
+            return
+        }
+        rateLimiter.close()
+        profiler.detachIndicator(ACTIVE_ASYNC_OPERATIONS)
+    }
+
+
+    private suspend fun <T> processTask(supplier: suspend () -> T): T {
+        try {
+            profiler.start("acquire_limit").use { limitAcquireTime ->
+                var limitAcquired = false
+                while (!limitAcquired) {
+                    if (state.get() == State.TERMINATE) {
+                        val message = "RateLimitedDispatcher [$name] received TERMINATE state while rate limiter acquiring"
+                        logger.error(message)
+                        throw RejectedExecutionException(message)
+                    }
+                    limitAcquired = rateLimiter.tryAcquire(3, ChronoUnit.SECONDS)
+                }
+                limitAcquireTime.stop()
+            }
+
+            // Since async operation may complete faster than Started method call
+            // it must be called before asynchronous operation started
+            asyncOperationStarted()
+
+            return profiler.profile(
+                    "supply_operation",
+                    Supplier { runBlocking { supplier.invoke() } }
```

Comment:
we cannot make blocking calls inside suspendable functions

look at aggregating-profiler-kotlin module